README.md
# fixelart

A simple utility to fix AI-generated pixelart. Read about how it came to be(and see examples) [here](https://szymon.codes/blog/fixing-ai-pixelart) or use the [playground](https:/szymon.codes/fixelart-playground)

|                                Input                                 |                                 Fixelified outputs                       |
| :------------------------------------------------------------------: | :------------------------------------------------------------------: |
| <img src="https://github.com/golota60/fixelart/blob/master/tests/assets/test-face.png" width="400"> | <img src="https://github.com/golota60/fixelart/blob/master/assets/images-preview4.png" width="400"> |
  
### Installation

Using yarn

```ts
yarn add fixelart
```

or npm

```ts
npm install fixelart
```

### Usage

The `fixImage` function accepts an object with the following structure

```ts
import { fixImage } from "fixelart";

export interface MinimumData {
  height: number;
  width: number;
  data: Buffer | Array<number>; // This is an array of RGBA values
}

export interface FixOptions {
  // Output pixel width
  outPixWidth: number;
  // Output pixel height
  outPixHeight: number;
  // Strategy to use
  strategy: StrategiesType;
  // Tolerance to be used for mode algorithms
  // e.g. for tolerance=1 rgba(1,1,1,1) would be "equal" to rgba(0,0,0,0) and rgba(2,2,2,2)
  tolerance?: number; // default = 1
  // Whether the output should be shrunk so that an output pixel is actually a pixel
  shrinkOutput?: boolean; // default = false
}

fixImage(png: MinimumData, options: FixOptions)
```

### Strategies

Below you can find all the existing strategies and what they do underneath the hood. PRs are open for more!

- `Strategies.MAJORITY` - take the color that occurs the most often in the block
- `Strategies.AVERAGE` - take the average([arithmetic mean](https://en.wikipedia.org/wiki/Arithmetic_mean)) of colors in the blocks
- `Strategies.HARMONIC` - take the [harmonic mean](https://en.wikipedia.org/wiki/Harmonic_mean) of colors
- `Strategies.GEOMETRIC` - take the [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean) of colors
- `Strategies.MIDRANGE` - take the [midrange](https://en.wikipedia.org/wiki/Mid-range) of colors
- `Strategies.QUADRATIC` - take the [quadratic mean](https://en.wikipedia.org/wiki/Root_mean_square) of colors
- `Strategies.CUBIC` - take the [cubic mean](https://en.wikipedia.org/wiki/Cubic_mean) of colors
- `Strategies.ALG(05|10|20|30|40|50|60|70|80|90)` - a mix; if a color is making up above X%, then take it, otherwise take the average

### Examples

Here's an example usage with `pngjs`.

```ts
import { fixImage } from "fixelart";
import fs from "fs";
import { PNGWithMetadata } from "pngjs";

export const loadPng = (fileName: string): PNGWithMetadata => {
  const file = fs.readFileSync(fileName);

  const png = PNG.sync.read(file);
  return png;
};

export const savePng = (png: PNGWithMetadata, pathToSave: string) => {
  let buff = PNG.sync.write(png);

  fs.writeFileSync(pathToSave, buff);
};

const png = loadPng(path);

const fixedImage = fixImage(png, { outPixWidth, outPixHeight, strategy });

savePng(fixedImage, "./output.png");
```

You could also use it pretty easily in the browser with Canvas Browser API(todo!) as demonstrated in the [demo](szymon.codes/fixelart-playground)

### Feature requests/Bugs

If you have found a bug or have any ideas for algorithms, or how the tool can improve don't hesitate to file an issue. Please provide the context though!

### Interactive examples

All the test images are located inside `tests/assets` folder

Before running anything, install all the dependencies by running `yarn`.

There are currently two examples

1. Generate all the algorithms for an images

This generates all the possible variations of your "fixed" pixelart for every rounding strategy(e.g. average, mean of the colors in the pixel).

Recipe:

```
yarn example-all-strategies <path_to_image> <pixel_width> <pixel_height>
```

example:

```
yarn example-all-strategies ./tests/assets/test-face.png 8 8
```

2. Generate all the test images for an algorithm

Recipe:

```
yarn example-all-images <strategy> <pixel_width> <pixel_height>
```

example:

```
yarn example-all-images MEAN 4 4
```

where

- `strategy` is one of the strategies
- `path_to_image` is the path to the source image
- `pixel_width` and `pixel_height` are values that indicate how much actual pixels are used in creating a pixel in output pixelart.
  e.g. if your source image is 1024x1024 and `pixel_width` and `pixel_height` are `4`, then the output image is going to be a 256-bit pixelart image.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.9](https://github.com/golota60/fixelart/compare/v0.0.8...v0.0.9) (2023-04-22)

### [0.0.8](https://github.com/golota60/fixelart/compare/v0.0.7...v0.0.8) (2023-04-22)

### [0.0.7](https://github.com/golota60/fixelart/compare/v0.0.6...v0.0.7) (2023-04-20)


### Bug Fixes

* geometric mean for big numbers ([7df3f03](https://github.com/golota60/fixelart/commit/7df3f03c6d9569ec5dc8cf63d364f4b071462f56))

### [0.0.6](https://github.com/golota60/fixelart/compare/v0.0.5...v0.0.6) (2023-04-20)


### Features

* add cubnic and quadratic means ([017fb60](https://github.com/golota60/fixelart/commit/017fb60d8d5b0c8729ea47853e27884b7de07480))
* add harmonic and geometric mean ([4e03768](https://github.com/golota60/fixelart/commit/4e0376868cf877626fb73abcca8cab14a2f4f7e6))
* add midrange alg ([402f88b](https://github.com/golota60/fixelart/commit/402f88b5b71ac095bf36c0ef3106be556e7bdac8))

### [0.0.5](https://github.com/golota60/fixelart/compare/v0.0.4...v0.0.5) (2023-04-19)


### Features

* add shrink option + better readme ([24fe168](https://github.com/golota60/fixelart/commit/24fe168d5f65c55996f8cba34e3eab363c4cd981))

### [0.0.4](https://github.com/golota60/fixelart/compare/v0.0.3...v0.0.4) (2023-04-19)


### Features

* support non-whole values ([05d435d](https://github.com/golota60/fixelart/commit/05d435dfd875c488ed5446e1c240e745d0bdcbe1))

### [0.0.3](https://github.com/golota60/fixelart/compare/v0.0.2...v0.0.3) (2023-04-18)


### Features

* some improvements ([78c1820](https://github.com/golota60/fixelart/commit/78c1820741606d82577a160a979df6b9a5880870))

### [0.0.2](https://github.com/golota60/fixelart/compare/v0.0.1...v0.0.2) (2023-04-17)

### 0.0.1 (2023-04-17)


### Features

* add majority alg ([df909e8](https://github.com/golota60/fixelart/commit/df909e80913f0facfbef77033d04b927b2fecfe4))
* add some algs ([055fe61](https://github.com/golota60/fixelart/commit/055fe61e10b6607bf6fa7ac280b3f20791d9b2b7))
* add some examples and general useful stuff ([fafc2e9](https://github.com/golota60/fixelart/commit/fafc2e9684b95cd826d3a33e30ed6294554c26d8))
* decouple libraries ([def8029](https://github.com/golota60/fixelart/commit/def8029fc3b0b4d5d09b75cccd1af3408b24424b))


tests/allImages.ts
import { StrategiesType, fixImage } from "../src";
import { loadPng, savePng } from "./testUtils";
import fs from "fs";

// Iterate over all assets for a given strategy
const generateAllImagesForStrategy = (
  outPixWidth: number,
  outPixHeight: number,
  strategy: StrategiesType
) => {
  const assetsData = fs.readdirSync("./assets");
  const files = fs.readdirSync(".");

  if (!files.includes("out")) {
    fs.mkdirSync("./out");
  }
  assetsData.forEach((fileName) => {
    const png = loadPng(fileName);

    const fixedImage = fixImage(png, { outPixWidth, outPixHeight, strategy });

    savePng(
      fixedImage,
      `./out/${fileName.split("/").at(-1)!.split(".png").at(-2)!}-out.png`
    );
  });
};

let args = process.argv.slice(2);

console.log("args passed:", args);

generateAllImagesForStrategy(
  Number(args[1]),
  Number(args[2]),
  args[0] as StrategiesType
);


tests/testUtils.ts
import { PNG, PNGWithMetadata } from "pngjs";
import fs from "fs";

/*
 * Loads a png using `pngjs`
 */
export const loadPng = (fileName: string): PNGWithMetadata => {
  const file = fs.readFileSync(fileName);

  const png = PNG.sync.read(file);
  return png;
};

/*
 * Saves a png using `pngjs`
 */
export const savePng = (png: PNGWithMetadata, pathToSave: string) => {
  let buff = PNG.sync.write(png);

  fs.writeFileSync(pathToSave, buff);
};


tests/allStrategies.ts
import { Strategies, fixImage } from "../src";
import { loadPng, savePng } from "./testUtils";
import fs from "fs";

// Iterate all strategires for an image
const generateAllStrategiesForImage = (
  outPixWidth: number,
  outPixHeight: number,
  path: string
) => {
  Object.values(Strategies).forEach((strategy) => {
    const png = loadPng(path);
    const files = fs.readdirSync(".");

    if (!files.includes("out")) {
      fs.mkdirSync("./out");
    }

    const fixedImage = fixImage(png, {
      outPixWidth,
      outPixHeight,
      strategy,
    });

    savePng(
      fixedImage,
      `./out/${path.split("/").at(-1)!.split(".png").at(-2)!}-${strategy}.png`
    );
  });
};

let args = process.argv.slice(2);

console.log("args passed:", args);

generateAllStrategiesForImage(Number(args[1]), Number(args[2]), args[0]);


out.js
let x = [
  [
    [
      [
        10,
        10,
        10,
        255
      ],
      [
        10,
        11,
        10,
        255
      ],
      [
        10,
        10,
        11,
        255
      ],
      [
        11,
        10,
        10,
        255
      ]
    ],
    [
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        31,
        255
      ],
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        30,
        255
      ]
    ],
    [
      [
        50,
        50,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ],
      [
        50,
        50,
        51,
        255
      ]
    ],
    [
      [
        11,
        10,
        10,
        255
      ],
      [
        10,
        10,
        10,
        255
      ],
      [
        10,
        10,
        10,
        255
      ],
      [
        11,
        10,
        10,
        255
      ]
    ]
  ],
  [
    [
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        30,
        255
      ],
      [
        31,
        30,
        30,
        255
      ],
      [
        30,
        31,
        30,
        255
      ]
    ],
    [
      [
        50,
        50,
        51,
        255
      ],
      [
        51,
        50,
        50,
        255
      ],
      [
        50,
        51,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ]
    ],
    [
      [
        10,
        10,
        10,
        255
      ],
      [
        10,
        10,
        10,
        255
      ],
      [
        10,
        10,
        11,
        255
      ],
      [
        11,
        10,
        10,
        255
      ]
    ],
    [
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        31,
        30,
        255
      ]
    ]
  ],
  [
    [
      [
        50,
        50,
        50,
        255
      ],
      [
        51,
        50,
        50,
        255
      ],
      [
        50,
        51,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ]
    ],
    [
      [
        11,
        10,
        10,
        255
      ],
      [
        10,
        11,
        10,
        255
      ],
      [
        10,
        10,
        11,
        255
      ],
      [
        11,
        10,
        10,
        255
      ]
    ],
    [
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        31,
        255
      ],
      [
        30,
        30,
        30,
        255
      ],
      [
        30,
        30,
        30,
        255
      ]
    ],
    [
      [
        50,
        50,
        51,
        255
      ],
      [
        51,
        50,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ],
      [
        50,
        50,
        50,
        255
      ]
    ]
  ],
  [
    [
      [
        70,
        70,
        70,
        255
      ],
      [
        70,
        71,
        70,
        255
      ],
      [
        70,
        70,
        70,
        255
      ],
      [
        70,
        70,
        70,
        255
      ]
    ],
    [
      [
        90,
        91,
        90,
        255
      ],
      [
        90,
        90,
        91,
        255
      ],
      [
        90,
        90,
        90,
        255
      ],
      [
        90,
        91,
        90,
        255
      ]
    ],
    [
      [
        110,
        110,
        110,
        255
      ],
      [
        111,
        110,
        110,
        255
      ],
      [
        110,
        111,
        110,
        255
      ],
      [
        110,
        110,
        110,
        255
      ]
    ],
    [
      [
        70,
        70,
        70,
        255
      ],
      [
        70,
        71,
        70,
        255
      ],
      [
        70,
        70,
        71,
        255
      ],
      [
        71,
        70,
        70,
        255
      ]
    ]
  ],
  [
    [
      [
        90,
        90,
        90,
        255
      ],
      [
        90,
        90,
        91,
        255
      ],
      [
        91,
        90,
        90,
        255
      ],
      [
        90,
        91,
        90,
        255
      ]
    ],
    [
      [
        110,
        110,
        111,
        255
      ],
      [
        110,
        110,
        110,
        255
      ],
      [
        110,
        110,
        110,
        255
      ],
      [
        110,
        110,
        111,
        255
      ]
    ],
    [
      [
        71,
        70,
        70,
        255
      ],
      [
        70,
        70,
        70,
        255
      ],
      [
        70,
        70,
        71,
        255
      ],
      [
        71,
        70,
        70,
        255
      ]
    ],
    [
      [
        90,
        91,
        90,
        255
      ],
      [
        90,
        90,
        91,
        255
      ],
      [
        90,
        90,
        90,
        255
      ],
      [
        90,
        91,
        90,
        255
      ]
    ]
  ],
  [
    [
      [
        110,
        110,
        111,
        255
      ],
      [
        111,
        110,
        110,
        255
      ],
      [
        110,
        110,
        110,
        255
      ],
      [
        110,
        110,
        110,
        255
      ]
    ],
    [
      [
        71,
        70,
        70,
        255
      ],
      [
        70,
        71,
        70,
        255
      ],
      [
        70,
        70,
        70,
        255
      ],
      [
        71,
        70,
        70,
        255
      ]
    ],
    [
      [
        90,
        90,
        90,
        255
      ],
      [
        90,
        90,
        90,
        255
      ],
      [
        91,
        90,
        90,
        255
      ],
      [
        90,
        90,
        90,
        255
      ]
    ],
    [
      [
        110,
        110,
        111,
        255
      ],
      [
        111,
        110,
        110,
        255
      ],
      [
        110,
        111,
        110,
        255
      ],
      [
        110,
        110,
        110,
        255
      ]
    ]
  ],
  [
    [
      [
        130,
        130,
        130,
        255
      ],
      [
        130,
        131,
        130,
        255
      ],
      [
        130,
        130,
        130,
        255
      ],
      [
        131,
        130,
        130,
        255
      ]
    ],
    [
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        150,
        151,
        255
      ],
      [
        151,
        150,
        150,
        255
      ],
      [
        150,
        151,
        150,
        255
      ]
    ],
    [
      [
        170,
        170,
        171,
        255
      ],
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ]
    ],
    [
      [
        131,
        130,
        130,
        255
      ],
      [
        130,
        130,
        130,
        255
      ],
      [
        130,
        130,
        131,
        255
      ],
      [
        131,
        130,
        130,
        255
      ]
    ]
  ],
  [
    [
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ]
    ],
    [
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        171,
        255
      ]
    ],
    [
      [
        131,
        130,
        130,
        255
      ],
      [
        130,
        130,
        130,
        255
      ],
      [
        130,
        130,
        130,
        255
      ],
      [
        130,
        130,
        130,
        255
      ]
    ],
    [
      [
        150,
        151,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ],
      [
        151,
        150,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ]
    ]
  ],
  [
    [
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ],
      [
        170,
        171,
        170,
        255
      ],
      [
        170,
        170,
        170,
        255
      ]
    ],
    [
      [
        131,
        130,
        130,
        255
      ],
      [
        130,
        131,
        130,
        255
      ],
      [
        130,
        130,
        131,
        255
      ],
      [
        131,
        130,
        130,
        255
      ]
    ],
    [
      [
        150,
        151,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        150,
        150,
        255
      ],
      [
        150,
        151,
        150,
        255
      ]
    ],
    [
      [
        170,
        170,
        170,
        255
      ],
      [
        171,
        170,
        170,
        255
      ],
      [
        170,
        171,
        170,
        255
      ],
      [
        170,
        170,
        171,
        255
      ]
    ]
  ]
]
Done in 4.35s.


lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const pngjs_1 = require("pngjs");
const loadPng = (fileName) => {
    //const file =
    const file = fs_1.default.readFileSync(fileName);
    const png = pngjs_1.PNG.sync.read(file);
    return png;
};
const savePng = (png, pathToSave) => {
    let buff = pngjs_1.PNG.sync.write(png);
    fs_1.default.writeFileSync(pathToSave, buff);
};
const accumulateColors = (pixels, i) => pixels.reduce((acc, item) => {
    return acc + item[i];
}, 0);
const getMeanOfColors = (pixels) => {
    const redAccumulated = accumulateColors(pixels, 0);
    const red = redAccumulated / pixels.length;
    const greenAccumulated = accumulateColors(pixels, 1);
    const green = greenAccumulated / pixels.length;
    const blueAccumulated = accumulateColors(pixels, 2);
    const blue = blueAccumulated / pixels.length;
    // TODO: Rethink whether alpha should be in this
    const alphaAccumulated = accumulateColors(pixels, 3);
    const alpha = alphaAccumulated / pixels.length;
    return [red, green, blue, alpha];
};
// input: a jpeg
// algorithm ideas:
//  - take the most prevalent color and make it the main
//  - take the mean of colors
//  - in between - e.g. if a color makes up less than <15% ignore it, otherwise add it to the mean.
//
//(some grid shifting might be needed e.g if the first row is 1.5 pixel high instead of 1)
//(auto-detect pixel size?)
//
//  output: a fixed jpeg
//
//
const Strategies = Object.freeze({
    MAJORITY: "majority",
    MEAN: "mean",
    MAJORITYWITHIGNORE: "majorityWithIgnore",
});
const fix = (outPixWidth, outPixHeight, strategy = Strategies.MAJORITYWITHIGNORE) => {
    const png = loadPng("./testgray.png");
    // alg here
    const imageHeight = png.height;
    const imageWidth = png.width;
    let imageData = png.data;
    const pixBlockSizeWidth = Number((imageWidth / outPixWidth).toFixed(0));
    const pixBlockSizeHeight = Number((imageHeight / outPixHeight).toFixed(0));
    console.log({ pixBlockSizeHeight, pixBlockSizeWidth });
    if (pixBlockSizeWidth !== pixBlockSizeHeight) {
        throw new Error(`Different pixel values for width and height. Pix block height: ${pixBlockSizeHeight}. Pix block width: ${pixBlockSizeWidth}`);
    }
    console.log({ imageWidth, imageHeight });
    let blocks = [];
    for (let hI = 0; hI < imageHeight; hI++) {
        for (let wI = 0; wI < imageWidth; wI++) {
            // go through all pixels per row
            const idStart = imageWidth * hI + wI;
            const idx = idStart << 2;
            //console.log(imageData[idx], idx, idStart, hI, wI, rows.length);
            const currentPixel = [
                imageData[idx],
                imageData[idx + 1],
                imageData[idx + 2],
                imageData[idx + 3],
            ];
            // Index in the current row
            const colIndex = wI % outPixWidth;
            const rowIndex = hI % outPixHeight;
            // this means we're at the row level of a new block
            if (rowIndex === 0 && colIndex === 0) {
                blocks.push([currentPixel]);
            }
            else {
                // Push to a corresponding block
                // Logic is - take the amount of blocks "over the current block"( Math.floor(hI / outPixHeight) * (outPixWidth - 1) )
                // and add the current column in the row we're currently traversing
                blocks[Math.floor(hI / outPixHeight) * (outPixWidth - 1) +
                    Math.floor(wI / outPixWidth)].push(currentPixel);
            }
        }
    }
    console.log(blocks);
    // Go through the blocks and mutate accordingly
    for (let bI = 0; bI < blocks.length; bI++) {
        const block = blocks[bI];
        switch (strategy) {
            case Strategies.MEAN:
                const mean = getMeanOfColors(block);
                console.log(mean);
                blocks[bI] = new Array(block.length).fill(mean);
                break;
            case Strategies.MAJORITY:
                break;
            case Strategies.MAJORITYWITHIGNORE:
                break;
        }
    }
    const flatmapped = blocks.flatMap((e) => e).flatMap((e) => e);
    for (let i = 0; i < imageData.length; i++) {
        imageData[i] = flatmapped[i];
    }
    console.log(flatmapped.length);
    console.log(imageData.length);
    savePng(png, "./test1out.png");
};
fix(4, 4, Strategies.MEAN);


src/utils.ts
// TODO: Add a 'shift' option? that would "shift" the start by N pixels

// size:4 -> r,g,b,a
export type Pixel = Array<number>;
// block - this should end up as lenth of <outPixWidth>*<outPixHeight>
export type Block = Array<Pixel>;

export interface MinimumData {
  height: number;
  width: number;
  // For our usage, it doesn't matter whether it's a buffer or an array
  data: Buffer | Array<number>;
}

export const Strategies = Object.freeze({
  // take the color that takes the majority of the block
  MAJORITY: "majority",
  // just take the mean out of colors in the current block
  AVERAGE: "average",
  // harmonic mean
  HARMONIC: "harmonic",
  // geometric mean
  GEOMETRIC: "geometric",
  // midrange
  MIDRANGE: "midrange",
  // quadratic mean
  QUADRATIC: "quadratic",
  // cubic mean
  CUBIC: "cubic",
  // take the color only if it is present for over X% of the image, otherwise take average
  ALG05: 0.05,
  ALG10: 0.1,
  ALG20: 0.2,
  ALG30: 0.3,
  ALG40: 0.4,
  ALG50: 0.5,
  ALG60: 0.6,
  ALG70: 0.7,
  ALG80: 0.8,
  ALG90: 0.9,
} as const);

export type StrategiesType = (typeof Strategies)[keyof typeof Strategies];

export interface FixOptions {
  outPixWidth: number;
  outPixHeight: number;
  strategy: StrategiesType;
  tolerance?: number;
  shrinkOutput?: boolean;
}

const accumulateColors = (pixels: Array<Pixel>, i: number) =>
  pixels.map((e) => e[i]);

const sumColors = (pixels: Array<Pixel>, i: number) =>
  pixels.reduce((acc, item) => {
    return acc + item[i];
  }, 0);

const multiplyColors = (pixels: Array<Pixel>, i: number) =>
  pixels.reduce((acc, item) => {
    return acc * item[i];
  }, 1);

export const getAverageOfColors = (pixels: Array<Pixel>): Pixel => {
  const redAccumulated = sumColors(pixels, 0);
  const red = redAccumulated / pixels.length;

  const greenAccumulated = sumColors(pixels, 1);
  const green = greenAccumulated / pixels.length;

  const blueAccumulated = sumColors(pixels, 2);
  const blue = blueAccumulated / pixels.length;

  // TODO: Rethink whether alpha should be in this
  const alphaAccumulated = sumColors(pixels, 3);
  const alpha = alphaAccumulated / pixels.length;

  return [
    Math.round(red),
    Math.round(green),
    Math.round(blue),
    Math.round(alpha),
  ];
};

const getHarmonicMean = (numsToAvg: Array<number>) => {
  const sumOfNegativePowers = numsToAvg
    .map((num) => 1 / num)
    .reduce((acc, num) => acc + num, 0);
  return Math.round(numsToAvg.length / sumOfNegativePowers);
};

export const getHarmonicMeanOfColors = (pixels: Array<Pixel>) => {
  const redAccumulated = accumulateColors(pixels, 0);
  const red = getHarmonicMean(redAccumulated);

  const greenAccumulated = accumulateColors(pixels, 1);
  const green = getHarmonicMean(greenAccumulated);

  const blueAccumulated = accumulateColors(pixels, 2);
  const blue = getHarmonicMean(blueAccumulated);

  const alphaAccumulated = accumulateColors(pixels, 3);
  const alpha = getHarmonicMean(alphaAccumulated);

  return [red, green, blue, alpha];
};

const getGeometricMean = (numsToAvg: Array<number>) => {
  const squareRooted = numsToAvg.map((e) => Math.pow(e, 1 / numsToAvg.length));

  const product = squareRooted.reduce((acc, item) => {
    const itemNormalized = item !== 0 ? item : 1;
    return acc * itemNormalized;
  }, 1);

  return Math.round(product);
};

export const getGeometricMeanOfColors = (pixels: Array<Pixel>) => {
  const redAccumulated = accumulateColors(pixels, 0);
  const red = getGeometricMean(redAccumulated);

  const greenAccumulated = accumulateColors(pixels, 1);
  const green = getGeometricMean(greenAccumulated);

  const blueAccumulated = accumulateColors(pixels, 2);
  const blue = getGeometricMean(blueAccumulated);

  const alphaAccumulated = accumulateColors(pixels, 3);
  const alpha = getGeometricMean(alphaAccumulated);

  return [red, green, blue, alpha];
};

const getMidrange = (numsToMidrange: Array<number>) => {
  const biggest = numsToMidrange.reduce(
    (nowBiggest, current) => (current > nowBiggest ? current : nowBiggest),
    0
  );
  const smallest = numsToMidrange.reduce(
    (nowSmallest, current) => (current < nowSmallest ? current : nowSmallest),
    255
  );
  return Math.round((biggest + smallest) / 2);
};

export const getMidrangeOfColors = (pixels: Array<Pixel>) => {
  const redAccumulated = accumulateColors(pixels, 0);
  const red = getMidrange(redAccumulated);

  const greenAccumulated = accumulateColors(pixels, 1);
  const green = getMidrange(greenAccumulated);

  const blueAccumulated = accumulateColors(pixels, 2);
  const blue = getMidrange(blueAccumulated);

  const alphaAccumulated = accumulateColors(pixels, 3);
  const alpha = getMidrange(alphaAccumulated);

  return [red, green, blue, alpha];
};

const getQuadraticMean = (numsToAvg: Array<number>) => {
  return Math.round(Math.sqrt(1 / numsToAvg.length) * Math.hypot(...numsToAvg));
};

export const getQuadraticMeanOfColors = (pixels: Array<Pixel>) => {
  const redAccumulated = accumulateColors(pixels, 0);
  const red = getQuadraticMean(redAccumulated);

  const greenAccumulated = accumulateColors(pixels, 1);
  const green = getQuadraticMean(greenAccumulated);

  const blueAccumulated = accumulateColors(pixels, 2);
  const blue = getQuadraticMean(blueAccumulated);

  const alphaAccumulated = accumulateColors(pixels, 3);
  const alpha = getQuadraticMean(alphaAccumulated);

  return [red, green, blue, alpha];
};

const getCubicMean = (numsToAvg: Array<number>) => {
  return Math.round(
    Math.cbrt(1 / numsToAvg.length) *
      Math.cbrt(numsToAvg.reduce((acc, elem) => acc + Math.pow(elem, 3), 0))
  );
};

export const getCubicMeanOfColors = (pixels: Array<Pixel>) => {
  const redAccumulated = accumulateColors(pixels, 0);
  const red = getCubicMean(redAccumulated);

  const greenAccumulated = accumulateColors(pixels, 1);
  const green = getCubicMean(greenAccumulated);

  const blueAccumulated = accumulateColors(pixels, 2);
  const blue = getCubicMean(blueAccumulated);

  const alphaAccumulated = accumulateColors(pixels, 3);
  const alpha = getCubicMean(alphaAccumulated);

  return [red, green, blue, alpha];
};

interface MajorityColorData {
  color: Pixel;
  occurences: number; // n of occurences
}

const compareWithTolerance = (
  val: number,
  val2: number,
  tolerance: number
): boolean => {
  return val + tolerance > val2 && val2 > val - tolerance;
};

const comparePixelPredicate = (
  pix: Pixel,
  pix2: Pixel,
  tolerance: number
): boolean => {
  return (
    compareWithTolerance(pix[0], pix2[0], tolerance) &&
    compareWithTolerance(pix[1], pix2[1], tolerance) &&
    compareWithTolerance(pix[2], pix2[2], tolerance) &&
    compareWithTolerance(pix[3], pix2[3], tolerance)
  );
};

export const getMajorityColor = (
  pixels: Array<Pixel>,
  tolerance = 1
): MajorityColorData => {
  // TODO: write this shit performantly instead of like this
  const colorsAndOccurences = pixels.reduce((acc, pixel) => {
    const found = acc.find((pixOccurData) =>
      comparePixelPredicate(pixOccurData.color, pixel, tolerance)
    );
    if (!!found) {
      const filteredOut = acc.filter(
        (pixOccurData) =>
          !comparePixelPredicate(pixOccurData.color, pixel, tolerance)
      );
      return [{ ...found, occurences: found.occurences + 1 }, ...filteredOut];
    }
    return [...acc, { color: pixel, occurences: 1 }];
  }, [] as Array<MajorityColorData>);

  const majority = colorsAndOccurences.reduce((acc, pixOccurData, index) => {
    if (index === 0) return pixOccurData;
    return acc.occurences > pixOccurData.occurences ? acc : pixOccurData;
  }, {} as MajorityColorData);

  return majority;
};

// input: a png

// algorithm ideas:
//  - take the most prevalent color and make it the main
//  - take the average/mean of colors
//  - in between - e.g. if a color makes up less than <15% ignore it, otherwise add it to the mean.
//
//(some grid shifting might be needed e.g if the first row is 1.5 pixel high instead of 1)
//(auto-detect pixel size?)
